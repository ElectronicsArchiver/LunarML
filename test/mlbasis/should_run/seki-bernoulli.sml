structure Rational :> sig
              eqtype rational
              val fromInt : Int.int -> rational
              val fromIntInf : IntInf.int -> rational
              val makeRational : IntInf.int * IntInf.int -> rational
              val numerator : rational -> IntInf.int
              val denominator : rational -> IntInf.int
              val + : rational * rational -> rational
              val - : rational * rational -> rational
              val * : rational * rational -> rational
              val / : rational * rational -> rational
              val abs : rational -> rational
              val ~ : rational -> rational
              val compare : rational * rational -> order
              val < : rational * rational -> bool
              val <= : rational * rational -> bool
              val > : rational * rational -> bool
              val >= : rational * rational -> bool
              val toString : rational -> string
          end = struct
(* invariant: den is strictly positive, gcd (num, den) = 1 *)
type rational = { num : IntInf.int, den : IntInf.int }
fun gcd (x, y) = gcdLoop (IntInf.abs x, IntInf.abs y)
and gcdLoop (x, y) = if y = IntInf.fromInt 0 then
                         x
                     else
                         gcdLoop (y, IntInf.rem (x, y))
fun makeRational (x, y) = if y = IntInf.fromInt 0 then
                              raise Div
                          else
                              let val g = gcd (x, y)
                              in if y < IntInf.fromInt 0 then
                                     { num = (~ x) div g, den = (~ y) div g }
                                 else
                                     { num = x div g, den = y div g }
                              end
fun fromInt x = { num = IntInf.fromInt x, den = IntInf.fromInt 1 }
fun fromIntInf (x : IntInf.int) = { num = x, den = IntInf.fromInt 1 }
val numerator = #num : rational -> IntInf.int
val denominator = #den : rational -> IntInf.int
fun add (x, y) = makeRational (numerator x * denominator y + numerator y * denominator x, denominator x * denominator y)
fun sub (x, y) = makeRational (numerator x * denominator y - numerator y * denominator x, denominator x * denominator y)
fun mul (x, y) = makeRational (numerator x * numerator y, denominator x * denominator y)
fun divide (x : rational, y : rational) = makeRational (numerator x * denominator y, denominator x * numerator y)
fun abs_ { num, den } = { num = IntInf.abs num, den = den : IntInf.int }
fun negate { num, den } = { num = IntInf.~ num, den = den : IntInf.int }
fun compare (x : rational, y : rational) = IntInf.compare (numerator x * denominator y, numerator y * denominator x)
fun x < y = IntInf.< (numerator x * denominator y, numerator y * denominator x)
fun x <= y = IntInf.<= (numerator x * denominator y, numerator y * denominator x)
fun x > y = IntInf.> (numerator x * denominator y, numerator y * denominator x)
fun x >= y = IntInf.>= (numerator x * denominator y, numerator y * denominator x)
fun toString x = IntInf.toString (numerator x) ^ "/" ^ IntInf.toString (denominator x)
val op + = add
val op - = sub
val op * = mul
val op / = divide
val abs = abs_
val ~ = negate
end

val limit = 100
val bern = Array.array (limit, Rational.fromInt 0)
val () = Array.update (bern, 0, Rational.fromInt 1)
val () = let fun loop (n, comb) = if n >= limit then
                                      ()
                                  else
                                      let val comb = Vector.tabulate (Vector.length comb + 1, fn k => if k = 0 orelse k = Vector.length comb then
                                                                                                          IntInf.fromInt 1
                                                                                                      else
                                                                                                          Vector.sub (comb, k - 1) + Vector.sub (comb, k)
                                                                     )
                                          val sum = let fun s (k, acc) = if k >= n then
                                                                             acc
                                                                         else
                                                                             s (k + 1, Rational.+ (acc, Rational.* (Rational.fromIntInf (Vector.sub (comb, k)), Array.sub (bern, k))))
                                                    in s (0, Rational.fromInt 0)
                                                    end
                                          val bern_n = Rational.~ (Rational./ (sum, Rational.fromInt (n + 1)))
                                      in Array.update (bern, n, bern_n)
                                       ; loop (n + 1, comb)
                                      end
         in loop (1, vector [IntInf.fromInt 1, IntInf.fromInt 1])
         end
val () = Array.appi (fn (i, x) => print ("B[" ^ Int.toString i ^ "]=" ^ Rational.toString x ^ "\n")) bern;
