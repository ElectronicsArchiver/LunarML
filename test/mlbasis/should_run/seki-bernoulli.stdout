B[0]=1/1
B[1]=~1/2
B[2]=1/6
B[3]=0/1
B[4]=~1/30
B[5]=0/1
B[6]=1/42
B[7]=0/1
B[8]=~1/30
B[9]=0/1
B[10]=5/66
B[11]=0/1
B[12]=~691/2730
B[13]=0/1
B[14]=7/6
B[15]=0/1
B[16]=~3617/510
B[17]=0/1
B[18]=43867/798
B[19]=0/1
B[20]=~174611/330
B[21]=0/1
B[22]=854513/138
B[23]=0/1
B[24]=~236364091/2730
B[25]=0/1
B[26]=8553103/6
B[27]=0/1
B[28]=~23749461029/870
B[29]=0/1
B[30]=8615841276005/14322
B[31]=0/1
B[32]=~7709321041217/510
B[33]=0/1
B[34]=2577687858367/6
B[35]=0/1
B[36]=~26315271553053477373/1919190
B[37]=0/1
B[38]=2929993913841559/6
B[39]=0/1
B[40]=~261082718496449122051/13530
B[41]=0/1
B[42]=1520097643918070802691/1806
B[43]=0/1
B[44]=~27833269579301024235023/690
B[45]=0/1
B[46]=596451111593912163277961/282
B[47]=0/1
B[48]=~5609403368997817686249127547/46410
B[49]=0/1
B[50]=495057205241079648212477525/66
B[51]=0/1
B[52]=~801165718135489957347924991853/1590
B[53]=0/1
B[54]=29149963634884862421418123812691/798
B[55]=0/1
B[56]=~2479392929313226753685415739663229/870
B[57]=0/1
B[58]=84483613348880041862046775994036021/354
B[59]=0/1
B[60]=~1215233140483755572040304994079820246041491/56786730
B[61]=0/1
B[62]=12300585434086858541953039857403386151/6
B[63]=0/1
B[64]=~106783830147866529886385444979142647942017/510
B[65]=0/1
B[66]=1472600022126335654051619428551932342241899101/64722
B[67]=0/1
B[68]=~78773130858718728141909149208474606244347001/30
B[69]=0/1
B[70]=1505381347333367003803076567377857208511438160235/4686
B[71]=0/1
B[72]=~5827954961669944110438277244641067365282488301844260429/140100870
B[73]=0/1
B[74]=34152417289221168014330073731472635186688307783087/6
B[75]=0/1
B[76]=~24655088825935372707687196040585199904365267828865801/30
B[77]=0/1
B[78]=414846365575400828295179035549542073492199375372400483487/3318
B[79]=0/1
B[80]=~4603784299479457646935574969019046849794257872751288919656867/230010
B[81]=0/1
B[82]=1677014149185145836823154509786269900207736027570253414881613/498
B[83]=0/1
B[84]=~2024576195935290360231131160111731009989917391198090877281083932477/3404310
B[85]=0/1
B[86]=660714619417678653573847847426261496277830686653388931761996983/6
B[87]=0/1
B[88]=~1311426488674017507995511424019311843345750275572028644296919890574047/61410
B[89]=0/1
B[90]=1179057279021082799884123351249215083775254949669647116231545215727922535/272118
B[91]=0/1
B[92]=~1295585948207537527989427828538576749659341483719435143023316326829946247/1410
B[93]=0/1
B[94]=1220813806579744469607301679413201203958508415202696621436215105284649447/6
B[95]=0/1
B[96]=~211600449597266513097597728109824233673043954389060234150638733420050668349987259/4501770
B[97]=0/1
B[98]=67908260672905495624051117546403605607342195728504487509073961249992947058239/6
B[99]=0/1
